### PR TITLE
[FLINK-14709] Allow outputting elements in user defined close method,…

### PR DESCRIPTION
**Pull request by dmvk**
## What is the purpose of the change

Allow outputting elements in user defined close method, executed by chained driver.

## Brief change log

- In BatchTask execution, close the output collector after chained tasks.
- In DataSourceTask execution, close the output collector after chained tasks.

## Verifying this change

This change added tests and can be verified as follows:

- Added tests for emitting output elements from UDF's close method.
- Verified by Beam's `validatesRunnerBatch` tests during BEAM-8608 implementation (backported to 1.9.1).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (kind of)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by dmvk** [comment]
cc @mxm @sunjincheng121

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java [hunk 1]
```diff
@@ -509,11 +509,14 @@ protected void run() throws Exception {
...
-			this.output.close();
-
 			// close all chained tasks letting them report failure
 			BatchTask.closeChainedTasks(this.chainedTasks, this);
+
+			// close the output collector
+			this.output.close();
+
 		}
+
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java]: Looks like an unrelated change. Please revert.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java [hunk 2]
```diff
@@ -208,12 +208,11 @@ public void invoke() throws Exception {
...
 			} // end for all input splits
 
-			// close the collector. if it is a chaining task collector, it will close its chained tasks
-			this.output.close();
-
 			// close all chained tasks letting them report failure
 			BatchTask.closeChainedTasks(this.chainedTasks, this);
 
+			// close the output collector
+			this.output.close();
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java]: +1 I think this order makes more sense.

**Comment by mxm** [comment]
@fhueske Could you have a look at this change? We want to output elements in the `close()` method. We need this for Beam bundle execution which may emit elements when it is finalized, e.g. when the operator gets closed.

**Comment by aljoscha** [comment]
This makes me wonder why the reduce/combine driver emit the final batch in `close()` and not `closeTask()`. I think it's a good change, though.

Logically it makes sense that we emit any last data in `closeTask()` and that `close()` is only for resource cleanup/dispose.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/chaining/ChainTaskTest.java [hunk 3]
```diff
@@ -174,7 +185,68 @@ public void testFailingMapTask() {
 			Assert.fail(e.getMessage());
 		}
 	}
-	
+
+	@Test
+	public void testOutputInCloseMethod_batchTask() {
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/chaining/ChainTaskTest.java]: The test methods should conform to standard java naming patterns, i.e. no underscore.

**Comment by aljoscha** [comment]
@dmvk Any update on my question?

**Comment by dmvk** [comment]
@aljoscha sry, I was on the vacation last week. I'll process your comments later today.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java (hunk 1 above)
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java]: fixed

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/chaining/ChainTaskTest.java (hunk 3 above)
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/chaining/ChainTaskTest.java]: fixed

**Comment by dmvk** [comment]
@aljoscha thanks for comments!

:+1: for moving emission of final batch to `closeTask` method.

`ChainedDriver` is a `Collector` (which is closable). When we `close` the head collector, all chained collectors get closed in a proper order and therefore we can emit more elements.

If we use `close` only for resource cleanup only, we can avoid code duplication in `cancelTask` (which need to perform resource cleanup as well), if we make sure to call `close` after both `closeTask` and `cancelTask` methods.

If we decide to do this, I think it should be done in separate JIRA / PR.

WDYT?

**Comment by aljoscha** [comment]
The reason I asked is because you did move the resource cleanup already so I thought we could do everything in one go, but you might be right.

Another thing: why does `close()` call `dispose(true)` while `cancelTask()` calls `dispose(false)`. This changes the behaviour of `cancelTask()` which ignored exceptions before. It seems to me that it should be the other way round, maybe?

**Comment by aljoscha** [comment]
By the way, I'm on the ASF slack workspace: https://s.apache.org/slack-invite. You can ping me there so that we can resolve reviews quicker. Sometimes it takes quite some time with the back-and-forth on github.

**Comment by dmvk** [comment]
You're right that the disposal exception handling is reversed `ChainedReduceCombineDriver`. Fixed, thanks!

**Comment by aljoscha** [comment]
Thanks for your work! I merged this.
